### PR TITLE
Refactor gcp tests to use pulumitest directly

### PR DESCRIPTION
**Pull request by VenelinMartinov**
The GCP upgrade tests were written in the early days of [pulumitest](https://github.com/pulumi/providertest/tree/main/pulumitest), so do not make use of some of the library features. This change refactors the tests to use it directly. This gets rid of some boilerplate code and allows direct access to the pulumitest options in tests.

It also attaches the provider directly which means we no longer have to keep rebuilding the provider ourselves to test it.

Additionally I've refactored TestCloudRunV2StateMigration which is an upgrade test with a source edit to use the library code instead of the code copied from the AWS provider.

### CODE REVIEW

**Review by iwahbe — APPROVED** [review]

File: provider/go.mod [hunk 1]
```diff
@@ -23,6 +23,8 @@ replace (
 	github.com/hashicorp/terraform-plugin-sdk/v2 => github.com/pulumi/terraform-plugin-sdk/v2 v2.0.0-20240520223432-0c0bf0d65f10
 	github.com/hashicorp/terraform-provider-google-beta => ../upstream
 	github.com/hashicorp/vault => github.com/hashicorp/vault v1.2.0
+	// Until we propagate this from the bridge, replace this here to fix a panic
```
Comment (guineveresaenger) [inline provider/go.mod:26]: We should link to an issue here.

File: provider/provider_test.go [hunk 2]
```diff
@@ -153,3 +78,27 @@ func pulumiTest(t *testing.T, dir string, opts ...opttest.Option) *pulumitest.Pu
...
 }
+
+func testUpgrade(t *testing.T, dir string, opts ...optproviderupgrade.PreviewProviderUpgradeOpt) auto.PreviewResult {
+	t.Helper()
+	if testing.Short() {
+		t.Skipf("Skipping in testing.Short() mode, assuming this is a CI run without cloud credentials")
+		return auto.PreviewResult{}
+	}
+
+	rpFactory := providers.ResourceProviderFactory(providerFactory)
```
Comment (guineveresaenger) [inline provider/provider_test.go:89]: non-blocking: what is the purpose of a `ResourceProviderFactory`? The [library docs](https://github.com/pulumi/providertest/blob/387f6b5565e7d34dc2314b9b12b8974894038a9e/providers/resourceProvider.go#L14) seem to describe a different function entirely.

More specifically to this PR: can we add a quick comment elaborating the purpose of this providerFactory (as opposed to `opttest.LocalProvider()`) - what does it allow the test framework to do?

**Review by guineveresaenger — APPROVED** [review]
non-blocking: would love to add a bit more context before merging for the sake of our future selves, e.g. let's link to the `pulumitest` module in the PR description so we disambiguate from `providertest`.

File: provider/provider_test.go [hunk 3]
```diff
@@ -40,96 +43,18 @@ func TestUpgradeCoverage(t *testing.T) {
...
-		 * We leave the schema blank. This will result in incorrect calls to
-		 * GetSchema, but otherwise does not effect the provider. It reduces the
-		 * time to test start by minutes.
-		 */
-		[]byte("{}"),
-	)(nil)
+	return pfbridge.MakeMuxedServer(ctx, info.Name, info, schemaBytes)(nil)
+}
+
+func providerServer(t *testing.T) pulumirpc.ResourceProviderServer {
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply provider/provider_test.go:56]: It's used in other tests

File: provider/provider_test.go (hunk 2 above)
Comment (VenelinMartinov) [reply provider/provider_test.go:90]: It allows us to use an in-process provider which automatically gets rebuilt for tests addressing one of your points from slack about tests not running against your latest changes.

**Pull request merged**